lang/wasm_alien_thread_runner.cc

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/reactor.hh>
+
+#include "lang/wasm.hh"
+#include "lang/wasm_alien_thread_runner.hh"
+
+using namespace std::chrono_literals;
+
+namespace wasm {
+
+std::optional<wasm_compile_task> task_queue::pop_front() {
+    if (_sem.try_acquire_for(10ms)) {
+        std::lock_guard l(_mut);
+        if (!_stopped) {
+            wasm_compile_task work_item = std::move(_pending.front());
+            _pending.pop();
+            return work_item;
+        }
+    }
+    return std::nullopt;
+}
+
+void task_queue::stop() {
+    std::lock_guard l(_mut);
+    _stopped = true;
+    while (_sem.try_acquire()) {
+        seastar::alien::run_on(seastar::engine().alien(), _pending.front().shard, [&done = _pending.front().done] {
+            done.set_exception(std::runtime_error("alien_thread_runner is stopping"));
+        });
+        _pending.pop();
+    }
+}
+
+void task_queue::push_back(wasm_compile_task work_item) {
+    std::lock_guard l(_mut);
+    if (_stopped) {
+        seastar::alien::run_on(seastar::engine().alien(), work_item.shard, [&done = work_item.done] {
+            done.set_exception(std::runtime_error("alien_thread_runner has stopped"));
+        });
+    } else if (_pending.size() >= QUEUE_SIZE) {
+        seastar::alien::run_on(seastar::engine().alien(), work_item.shard, [&done = work_item.done] {
+            done.set_exception(std::runtime_error("alien_thread_runner queue is full"));
+        });
+    } else {
+        _pending.push(work_item);
+        _sem.release();
+    }
+}
+
+alien_thread_runner::alien_thread_runner()
+    : thread([this] {
+        for (;;) {
+            auto work_item = pending_queue.pop_front();
+            if (work_item) {
+                work_item->func();
+            } else if (_stopping) {
+                break;
+            }
+        }
+    })
+{ }
+
+alien_thread_runner::~alien_thread_runner() {
+    thread.join();
```

Comment:
how about `assert(pending_queue.size() == 0`? if we don't call `stop()` in the dtor, we might need to make sure the user of `alien_thread_runner` class does this. or, we can do so in the dtor of `task_queue`.

Reply:
I moved the stop to the dtor in the rebase